### PR TITLE
Bootstrap Opam with OCaml 4.11.1

### PR DESCRIPTION
**Pull request by MisterDA**
This PR builds on https://github.com/ocaml/opam/pull/4234 to fix the little things that are missing and completely replaces 4.07.1 with 4.10.0 for bootstrapping.
Updating findlib to 1.8.1 is mandatory because 1.8.0 checks for the `-vmthread` flag of `ocamlc` which has been removed in 4.10.0.
EDIT: I removed the 'flexdll appveyor patch' in favor of using the latest flexdll commit available at the time of writing.
I have also updated and expanded Travis test matrix.

### CODE REVIEW

File: shell/bootstrap-ocaml.sh
Comment (dra27) [inline shell/bootstrap-ocaml.sh]: There is still the odd Solaris out there, and the default `sh` on still just-about support Solaris doesn't recognise `$()`. These need to stay as they are (the ones in the Windows-specific part don't matter).

**Comment by dra27** [comment]
You do actually need to `configure` the Windows build 😉 This is working for me:
```diff
diff --git a/shell/bootstrap-ocaml.sh b/shell/bootstrap-ocaml.sh
index 3cf90659..13aea4fd 100755
--- a/shell/bootstrap-ocaml.sh
+++ b/shell/bootstrap-ocaml.sh
@@ -51,12 +51,23 @@ PATH_PREPEND=
 LIB_PREPEND=
 INC_PREPEND=
 if [ -n "$1" -a -n "${COMSPEC}" -a -x "${COMSPEC}" ] ; then
+  case "$(uname -m)" in
+    'i686')
+      BUILD=i686-pc-cygwin
+    ;;
+    'x86_64')
+      BUILD=x86_64-pc-cygwin
+    ;;
+  esac
   case "$1" in
-    "mingw"|"mingw64")
-      BUILD=$1
+    "mingw")
+      HOST=i686-w64-mingw32
+    ;;
+    "mingw64")
+      HOST=x86_64-w64-mingw32
     ;;
     "msvc")
-      BUILD=$1
+      HOST=i686-pc-windows
       if ! command -v ml > /dev/null ; then
         eval `${ROOT}../../shell/msvs-detect --arch=x86`
         if [ -n "${MSVS_NAME}" ] ; then
@@ -67,7 +78,7 @@ if [ -n "$1" -a -n "${COMSPEC}" -a -x "${COMSPEC}" ] ; then
       fi
     ;;
     "msvc64")
-      BUILD=$1
+      HOST=x86_64-pc-windows
       if ! command -v ml64 > /dev/null ; then
         eval `${ROOT}../../shell/msvs-detect --arch=x64`
         if [ -n "${MSVS_NAME}" ] ; then
@@ -88,24 +99,24 @@ if [ -n "$1" -a -n "${COMSPEC}" -a -x "${COMSPEC}" ] ; then
       fi

       if [ ${TRY64} -eq 1 ] && command -v x86_64-w64-mingw32-gcc > /dev/null ; then
-        BUILD=mingw64
+        HOST=x86_64-w64-mingw32
       elif command -v i686-w64-mingw32-gcc > /dev/null ; then
-        BUILD=mingw
+        HOST=i686-w64-mingw32
       elif [ ${TRY64} -eq 1 ] && command -v ml64 > /dev/null ; then
-        BUILD=msvc64
+        HOST=x86_64-pc-windows
         PATH_PREPEND=`bash ${ROOT}../../shell/check_linker`
       elif command -v ml > /dev/null ; then
-        BUILD=msvc
+        HOST=i686-pc-windows
         PATH_PREPEND=`bash ${ROOT}../../shell/check_linker`
       else
         if [ ${TRY64} -eq 1 ] ; then
-          BUILD=msvc64
-          BUILD_ARCH=x64
+          HOST=x86_64-pc-windows
+          HOST_ARCH=x64
         else
-          BUILD=msvc
-          BUILD_ARCH=x86
+          HOST=i686-pc-windows
+          HOST_ARCH=x86
         fi
-        eval `${ROOT}../../shell/msvs-detect --arch=${BUILD_ARCH}`
+        eval `${ROOT}../../shell/msvs-detect --arch=${HOST_ARCH}`
         if [ -z "${MSVS_NAME}" ] ; then
           echo "No appropriate C compiler was found -- unable to build OCaml"
           exit 1
@@ -123,9 +134,8 @@ if [ -n "$1" -a -n "${COMSPEC}" -a -x "${COMSPEC}" ] ; then
   PREFIX=`cd .. ; pwd`/ocaml
   WINPREFIX=`echo ${PREFIX} | cygpath -f - -m`
   if [ ${GEN_CONFIG_ONLY} -eq 0 ] ; then
-    sed -e "s|^PREFIX=.*|PREFIX=${WINPREFIX}|" -e "s|/lib|/lib/ocaml|" config/Makefile.${BUILD} > config/Makefile
-    cp config/s-nt.h byterun/caml/s.h
-    cp config/m-nt.h byterun/caml/m.h
+    # --disable-ocamldoc can change to --disable-stdlib-manpages when bumped to 4.11
+    PATH="${PATH_PREPEND}${PREFIX}/bin:${PATH}" Lib="${LIB_PREPEND}${Lib}" Include="${INC_PREPEND}${Include}" ./configure --prefix "$WINPREFIX" --build=$BUILD --host=$HOST --disable-ocamldoc
   fi
   cd ..
   if [ ! -e ${ROOT}${FLEXDLL} ]; then
@@ -136,7 +146,9 @@ if [ -n "$1" -a -n "${COMSPEC}" -a -x "${COMSPEC}" ] ; then
     tar -xzf ${ROOT}../${FLEXDLL}
     rm -rf flexdll
     mv flexdll-* flexdll
-    PATH="${PATH_PREPEND}${PREFIX}/bin:${PATH}" Lib="${LIB_PREPEND}${Lib}" Include="${INC_PREPEND}${Include}" make flexdll world.opt install
+    PATH="${PATH_PREPEND}${PREFIX}/bin:${PATH}" Lib="${LIB_PREPEND}${Lib}" Include="${INC_PREPEND}${Include}" make -j flexdll
+    PATH="${PATH_PREPEND}${PREFIX}/bin:${PATH}" Lib="${LIB_PREPEND}${Lib}" Include="${INC_PREPEND}${Include}" make -j world.opt
+    PATH="${PATH_PREPEND}${PREFIX}/bin:${PATH}" Lib="${LIB_PREPEND}${Lib}" Include="${INC_PREPEND}${Include}" make install
     mkdir -p ../../src_ext
   fi
   OCAMLLIB=${WINPREFIX}/lib/ocaml
```

File: shell/bootstrap-ocaml.sh
Comment (MisterDA) [reply shell/bootstrap-ocaml.sh]: oh wow! I did not know that. They’re not using xpg4 or xpg6?

Comment (dra27) [reply shell/bootstrap-ocaml.sh]: I don't know the full detail - I just know that Solaris 10 uses an incredibly old `sh` by default (and IIRC a crazy `patch` too). I'm all for `$()` but for the most part, I tend to leave existing uses of backtick alone (you can still shoot yourself in the foot: cf. https://github.com/ocaml/opam/pull/3699#issuecomment-446939748)

Comment (MisterDA) [reply shell/bootstrap-ocaml.sh]: What I’m seeing is that Solaris provides multiples implementations of the tools, each following a standard.

> An application that wants to use standard-conforming utilities must set the PATH (sh(1) or ksh(1)) or path (csh(1)) environment variable to specify the directories listed below in the order specified to get the appropriate utilities:
>
> **POSIX.1–2001, SUSv3**
>
> - `/usr/xpg6/bin`
> - `/usr/xpg4/bin`
> - `/usr/bin`
> - directory containing binaries for your compiler
> - other directories containing binaries needed by the application

https://docs.oracle.com/cd/E86824_01/html/E54776/xpg-5.html

We *could* set the `PATH` to use the latest stuff :smile:

**Comment by MisterDA** [comment]
@dra27 there’s something weird in your patch. The line `mkdir -p ../../src_ext` is not present in the file in master, but it is not marked as added in your patch. Is it a left over from something else?

**Comment by dra27** [comment]
Oops, that patch’s against a different branch which supports building 4 bootstrap compilers! It should be sufficient just to ignore the `ROOT` variable!

**Comment by MisterDA** [comment]
@dra27 an idea why the msvc builds are failing?
And what's happening in Travis with the last failing builds?

**Comment by dra27** [comment]
This should unlock the Travis failure:
```diff
diff --git a/.travis-ci.sh b/.travis-ci.sh
index 9d85a845..f4621b91 100755
--- a/.travis-ci.sh
+++ b/.travis-ci.sh
@@ -192,6 +192,7 @@ EOF
           make world.opt
         fi
         make install
+        rm -rf "ocaml-$OCAML_VERSION.tar.gz"
         echo "LOCAL_OCAML_VERSION=$OCAML_VERSION" > ~/local/versions
         echo -en "travis_fold:end:ocaml\r"
       fi
```

**Comment by dra27** [comment]
The msvc failure is more subtle - I'm looking into it

**Comment by dra27** [comment]
OK, there's a path-blowout in `msvs-detect` - I'm just working on a fix. Hopefully the upstream maintainer for msvs-tools will merge it quickly :wink:

**Comment by dra27** [comment]
Sorry - that patch I gave you was wrong - it's the directory which needs removing, not the tarball.

**Comment by dra27** [comment]
i.e.

```diff
diff --git a/.travis-ci.sh b/.travis-ci.sh
index 9d85a845..73bc1670 100755
--- a/.travis-ci.sh
+++ b/.travis-ci.sh
@@ -192,6 +192,7 @@ EOF
           make world.opt
         fi
         make install
+        rm -rf "ocaml-$OCAML_VERSION"
         echo "LOCAL_OCAML_VERSION=$OCAML_VERSION" > ~/local/versions
         echo -en "travis_fold:end:ocaml\r"
       fi
```

**Comment by MisterDA** [comment]
I think that at this point we’re in the directory `"ocaml-$OCAML_VERSION"` (line 173) so we should `cd ..` before removing it.

**Comment by dra27** [comment]
Incidentally, and hopefully trivially, this should be 4.10.1

**Comment by rjbou** [comment]
@MisterDA Please cherry-pick 1b0f4bacae7fe64b23b802fe04ae75d5c203ce24 as it is reverted in master

**Comment by MisterDA** [comment]
Rebased against master. Some of the changes were integrated when transitioning from 4.07 to 4.09.1 in bootstrap, and some other fixes were integrated independently.

File: src_ext/Makefile
Comment (dra27) [inline src_ext/Makefile:18]: Eek - this is serious, as that should have been being caught by CI. I've pushed corrections to `src_ext/Makefile` and the consequential MD5 change to #4260

File: src_ext/Makefile
Comment (dra27) [inline src_ext/Makefile:14]:
```suggestion
URL_ocaml = https://caml.inria.fr/pub/distrib/ocaml-4.10/ocaml-4.10.0.tar.gz
```

Comment (dra27) [inline src_ext/Makefile:14]: although, given that we release the GitHub URLs to opam-repository, I think:
```suggestion
URL_ocaml = https://github.com/ocaml/ocaml/archive/4.10.0.tar.gz
```
is better, although you'll need the small change in https://github.com/dra27/opam/commit/755a760820fbd8ce331245e78a0170365cd5bca2#diff-a9c3fb4f36b2f2f2d0a5a6fc475ede42 (that change may be a good idea anyway as it makes testing a variant compiler easier)

File: .travis-ci.sh
Comment (MisterDA) [reply .travis-ci.sh:394]: Should the ref be moved forward for this PR?

Comment (dra27) [reply .travis-ci.sh:394]: Only if new packages are needed - 
```
dra27@summer:~/opam-repository$ git show 8c45759d4
commit 8c45759d4217eb0b3fbad41054f08a3f99fa9dd7 (HEAD -> master, upstream/master)
Merge: 4d326d82aa 0008b955ec
Author: Marcello Seri <mseri@users.noreply.github.com>
Date:   Tue Jun 30 16:38:14 2020 +0200

    Merge pull request #16733 from SGrondin/opam-publish-SZXX.0.1.0

    Package SZXX.0.1.0
```

Comment (dra27) [reply .travis-ci.sh:394]: (the purpose of the test is to ensure that the opam you just built is capable of performing an operation all the previous opams could do, so the idea is to change the operation as infrequently as possible)

File: src_ext/Makefile
Comment (MisterDA) [reply src_ext/Makefile]: Switched to the GitHub URL.

File: .travis-ci.sh
Comment (MisterDA) [reply .travis-ci.sh]: It seems to be recent enough.

**Comment by MisterDA** [comment]
Rebase on master (2.1.0-alpha3) and test OCaml 4.10.1+rc1.

**Review by dra27 — APPROVED** [review]
I just pushed one additional commit to action a comment in `shell/bootstrap_ocaml.sh`. Good to merge once CI comes back confirming that I didn't break the script!

**Comment by AltGr** [comment]
Thanks!

**Pull request merged**
